diff -rupN dejagnu-1.5.orig/lib/framework.exp dejagnu-1.5/lib/framework.exp
--- dejagnu-1.5.orig/lib/framework.exp	2011-03-03 12:58:56.000000000 +0300
+++ dejagnu-1.5/lib/framework.exp	2014-10-11 21:51:26.000000000 +0400
@@ -131,6 +131,9 @@ proc is_remote { board } {
     global host_board
     global target_list
 
+    # FLASCC HACK
+    return 0
+
     verbose "calling is_remote $board" 3
     # Remove any target variant specifications from the name.
     set board [lindex [split $board "/"] 0]
